{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **トランスポーター配列と基質のペアデータセットを構築するノートブック**\n",
    "(参考) [SPOT論文](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3002807) [ソースコード](https://github.com/AlexanderKroll/SPOT/blob/main/code/preprocessing/A3%20-%20Extract%20Transporter-substrate%20pairs%20from%20UniProt.ipynb)\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **0. 目的**\n",
    "\n",
    "このノートブックでは、UniProt、Rhea、ChEBIのデータを統合し、トランスポータータンパク質配列とその基質のペアデータセットを構築します。\n",
    "具体的な手順は以下の通りです。\n",
    "\n",
    "-  **UniProtデータの抽出:**\n",
    "    *   UniProtデータベースから、特定の条件（長さ、フラグメントの有無、GOアノテーション、触媒活性）を満たすタンパク質配列を抽出します。\n",
    "-  **基質の特定:**\n",
    "    *   Rheaデータベースから、Rhea IDとChEBI IDの対応関係を取得します。\n",
    "    *   ChEBIデータベースから、ChEBI IDと基質名の対応関係を取得します。\n",
    "-  **タンパク質配列と基質の対応付け:**\n",
    "    *   UniProtデータとRheaデータをRhea IDで結合します。\n",
    "    *   RheaデータとChEBIデータをChEBI IDで結合します。\n",
    "    *   結果として得られたデータから、タンパク質配列と基質名のペアを作成します。\n",
    "-  **データセットの作成と保存:**\n",
    "    *   データセットをトレーニング用とテスト用に分割します。\n",
    "    *   各データセットをJSON形式で保存し、モデルの学習に利用可能な形にします。\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **1. データの取得**\n",
    "\n",
    "解析に必要なデータは以下のリンクから取得します。\n",
    "\n",
    "-   **RheaとChEBI IDの対応表** (`rhea-reactions.txt`)\n",
    "    -   [Rhea FTP](https://ftp.expasy.org/databases/rhea/txt/)\n",
    "\n",
    "-   **ChEBI IDとChEBI名の対応表** (`compounds.tsv`)\n",
    "    -   [ChEBI FTP](https://ftp.ebi.ac.uk/pub/databases/chebi/Flat_file_tab_delimited/)\n",
    "\n",
    "-   **UniProtから特定条件を満たすデータセット** (`uniprotkb_length_50_TO_1024_AND_fragmen_2025_03_04.tsv`)\n",
    "    -  **条件**\n",
    "        -   長さ：50〜1024\n",
    "        -   フラグメント：なし\n",
    "        -   GOアノテーション：`go:0005215`（トランスポーター活性）\n",
    "        -   触媒活性：含む\n",
    "    -   クエリ：\n",
    "        ```\n",
    "        (length:[50 TO 1024]) AND (fragment:false) AND (go:0005215) AND (cc_catalytic_activity:*)\n",
    "        ```\n",
    "    -   **UniProtでクエリを実行してtsv形式でダウンロード**\n",
    "        ![alt text](image.png)\n",
    "\n",
    "**データ構造**\n",
    "\n",
    "ダウンロードした3つのファイルは`dataset`ディレクトリに配置してください。\n",
    "\n",
    "```bash\n",
    "dataset/\n",
    "    ├── compounds.tsv  # ChEBIデータ\n",
    "    ├── rhea-reactions.txt  # Rheaデータ\n",
    "    └── uniprotkb_length_50_TO_1024_AND_fragmen_2025_03_04.tsv  # UniProtデータ\n",
    "```\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **2. 必要なライブラリのインポート**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from os.path import join\n",
    "import subprocess\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "\n",
    "# 警告を無視\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## **3. UniProtデータの処理**\n",
    "\n",
    "### **(1) UniProtデータの読み込み**\n",
    "\n",
    "-   タブ区切り（TSV）で保存されているUniProtデータを読み込みます。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Catalytic activity</th>\n",
       "      <th>Gene Ontology (molecular function)</th>\n",
       "      <th>Annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...</td>\n",
       "      <td>ABC-type transporter activity [GO:0140359]; AT...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A024RA31</td>\n",
       "      <td>MASEFKKKLFWRAVVAEFLATTLFVFISIGSALGFKYPVGNNQTAV...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=CO2(out) = CO2(in...</td>\n",
       "      <td>ephrin receptor binding [GO:0046875]; potassiu...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A034VPM6</td>\n",
       "      <td>MLYCPPNVTFSQIWINNGISHCFMDTIGNGVASGFMLLMGTVQLLM...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...</td>\n",
       "      <td>ABC-type heme transporter activity [GO:0015439...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A060X8Q2</td>\n",
       "      <td>MVVVQSYCEASASISEAWLEGGISPCFYFTLVPTVLLTLSFFFGAF...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...</td>\n",
       "      <td>ABC-type heme transporter activity [GO:0015439...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A060XRG5</td>\n",
       "      <td>MAENMEGKVSNGEAHLDEPVANGLRHKKCSNETTVEKIKRLVMANL...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=D-serine(in) + L-...</td>\n",
       "      <td>antiporter activity [GO:0015297]; L-aspartate ...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466198</th>\n",
       "      <td>X8JNV3</td>\n",
       "      <td>MSSTIHTPPTRPQNVLPESPGIQGRSSYIYSDTVQTRQDSLEISEK...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=H2O(in) = H2O(out...</td>\n",
       "      <td>glycerol channel activity [GO:0015254]; water ...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466199</th>\n",
       "      <td>X8JR86</td>\n",
       "      <td>MGSIDRRNVISAVFVGFASLGGFIMGFDTSVISGVKELPAWKQRFG...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=myo-inositol(out)...</td>\n",
       "      <td>carbohydrate:proton symporter activity [GO:000...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466200</th>\n",
       "      <td>X8JTZ6</td>\n",
       "      <td>MAAGWLADRIGRKRTIQAGSLVAILGCSLQTGAQNINFLLVGRVIA...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=myo-inositol(out)...</td>\n",
       "      <td>carbohydrate:proton symporter activity [GO:000...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466201</th>\n",
       "      <td>X8JVQ7</td>\n",
       "      <td>MGKGYSPRIAANPYIVGSFACIGGGLFGLDISSMSGVLSNPSYLRV...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=myo-inositol(out)...</td>\n",
       "      <td>carbohydrate:proton symporter activity [GO:000...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466202</th>\n",
       "      <td>X8JWV1</td>\n",
       "      <td>MRLPRGSRVTGVLIRLAPWVPTFIAQEASGGDSEVKRAPPTAQAFR...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=myo-inositol(out)...</td>\n",
       "      <td>carbohydrate:proton symporter activity [GO:000...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1466203 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Entry                                           Sequence  \\\n",
       "0        A0A015JYU2  MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...   \n",
       "1        A0A024RA31  MASEFKKKLFWRAVVAEFLATTLFVFISIGSALGFKYPVGNNQTAV...   \n",
       "2        A0A034VPM6  MLYCPPNVTFSQIWINNGISHCFMDTIGNGVASGFMLLMGTVQLLM...   \n",
       "3        A0A060X8Q2  MVVVQSYCEASASISEAWLEGGISPCFYFTLVPTVLLTLSFFFGAF...   \n",
       "4        A0A060XRG5  MAENMEGKVSNGEAHLDEPVANGLRHKKCSNETTVEKIKRLVMANL...   \n",
       "...             ...                                                ...   \n",
       "1466198      X8JNV3  MSSTIHTPPTRPQNVLPESPGIQGRSSYIYSDTVQTRQDSLEISEK...   \n",
       "1466199      X8JR86  MGSIDRRNVISAVFVGFASLGGFIMGFDTSVISGVKELPAWKQRFG...   \n",
       "1466200      X8JTZ6  MAAGWLADRIGRKRTIQAGSLVAILGCSLQTGAQNINFLLVGRVIA...   \n",
       "1466201      X8JVQ7  MGKGYSPRIAANPYIVGSFACIGGGLFGLDISSMSGVLSNPSYLRV...   \n",
       "1466202      X8JWV1  MRLPRGSRVTGVLIRLAPWVPTFIAQEASGGDSEVKRAPPTAQAFR...   \n",
       "\n",
       "                                        Catalytic activity  \\\n",
       "0        CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...   \n",
       "1        CATALYTIC ACTIVITY: Reaction=CO2(out) = CO2(in...   \n",
       "2        CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...   \n",
       "3        CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...   \n",
       "4        CATALYTIC ACTIVITY: Reaction=D-serine(in) + L-...   \n",
       "...                                                    ...   \n",
       "1466198  CATALYTIC ACTIVITY: Reaction=H2O(in) = H2O(out...   \n",
       "1466199  CATALYTIC ACTIVITY: Reaction=myo-inositol(out)...   \n",
       "1466200  CATALYTIC ACTIVITY: Reaction=myo-inositol(out)...   \n",
       "1466201  CATALYTIC ACTIVITY: Reaction=myo-inositol(out)...   \n",
       "1466202  CATALYTIC ACTIVITY: Reaction=myo-inositol(out)...   \n",
       "\n",
       "                        Gene Ontology (molecular function)  Annotation  \n",
       "0        ABC-type transporter activity [GO:0140359]; AT...         5.0  \n",
       "1        ephrin receptor binding [GO:0046875]; potassiu...         5.0  \n",
       "2        ABC-type heme transporter activity [GO:0015439...         5.0  \n",
       "3        ABC-type heme transporter activity [GO:0015439...         5.0  \n",
       "4        antiporter activity [GO:0015297]; L-aspartate ...         5.0  \n",
       "...                                                    ...         ...  \n",
       "1466198  glycerol channel activity [GO:0015254]; water ...         1.0  \n",
       "1466199  carbohydrate:proton symporter activity [GO:000...         1.0  \n",
       "1466200  carbohydrate:proton symporter activity [GO:000...         1.0  \n",
       "1466201  carbohydrate:proton symporter activity [GO:000...         1.0  \n",
       "1466202  carbohydrate:proton symporter activity [GO:000...         1.0  \n",
       "\n",
       "[1466203 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniprot = pd.read_csv('dataset/uniprotkb_length_50_TO_1024_AND_fragmen_2025_03_04.tsv', sep='\\t')\n",
    "# df_uniprotを表示\n",
    "df_uniprot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **(2) Annotationスコアが3.0以上のものを抽出**\n",
    "\n",
    "-   Annotationスコアはタンパク質の注釈の信頼度を表します。\n",
    "-   信頼度の高いデータのみを対象にするため、スコアが3.0以上のものを抽出します。\n",
    "\n",
    "### **(4) 標準アミノ酸のみのシーケンスにフィルタリング**\n",
    "\n",
    "-   標準アミノ酸以外の文字が含まれている場合は除外されます。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Catalytic activity</th>\n",
       "      <th>Gene Ontology (molecular function)</th>\n",
       "      <th>Annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...</td>\n",
       "      <td>ABC-type transporter activity [GO:0140359]; AT...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A024RA31</td>\n",
       "      <td>MASEFKKKLFWRAVVAEFLATTLFVFISIGSALGFKYPVGNNQTAV...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=CO2(out) = CO2(in...</td>\n",
       "      <td>ephrin receptor binding [GO:0046875]; potassiu...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A034VPM6</td>\n",
       "      <td>MLYCPPNVTFSQIWINNGISHCFMDTIGNGVASGFMLLMGTVQLLM...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...</td>\n",
       "      <td>ABC-type heme transporter activity [GO:0015439...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A060X8Q2</td>\n",
       "      <td>MVVVQSYCEASASISEAWLEGGISPCFYFTLVPTVLLTLSFFFGAF...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...</td>\n",
       "      <td>ABC-type heme transporter activity [GO:0015439...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A060XRG5</td>\n",
       "      <td>MAENMEGKVSNGEAHLDEPVANGLRHKKCSNETTVEKIKRLVMANL...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=D-serine(in) + L-...</td>\n",
       "      <td>antiporter activity [GO:0015297]; L-aspartate ...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520258</th>\n",
       "      <td>Z9JMX9</td>\n",
       "      <td>MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=K(+)(out) + ATP +...</td>\n",
       "      <td>ATP binding [GO:0005524]; ATP hydrolysis activ...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520259</th>\n",
       "      <td>Z9JTY6</td>\n",
       "      <td>MTTQETRKRSTAVVRPDEERPLGDIGARVRAGVAAWTRSPALDFYA...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=[GlcNAc-(1-&gt;4)-Mu...</td>\n",
       "      <td>glycosyltransferase activity [GO:0016757]; lip...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520260</th>\n",
       "      <td>Z9JU65</td>\n",
       "      <td>MNPDTITLSAGSLTTVAVIVVIALIAVAMALRFRAEVLRAATGTPA...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=diphosphate + H2O...</td>\n",
       "      <td>diphosphate hydrolysis-driven proton transmemb...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520261</th>\n",
       "      <td>Z9JXE3</td>\n",
       "      <td>MAELTIRPEDIRHALDTAVSTYQAGSTDREEIGRVTESADGIARVE...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=ATP + H2O + 4 H(+...</td>\n",
       "      <td>ADP binding [GO:0043531]; ATP binding [GO:0005...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520262</th>\n",
       "      <td>Z9JZ28</td>\n",
       "      <td>MTTTAPETVTATGRVARVTGAVVDIEFPPGNIPAMYNALTTVIEAT...</td>\n",
       "      <td>CATALYTIC ACTIVITY: Reaction=ATP + H2O + 4 H(+...</td>\n",
       "      <td>ATP binding [GO:0005524]; ATP hydrolysis activ...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>405820 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry                                           Sequence  \\\n",
       "0       A0A015JYU2  MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...   \n",
       "1       A0A024RA31  MASEFKKKLFWRAVVAEFLATTLFVFISIGSALGFKYPVGNNQTAV...   \n",
       "2       A0A034VPM6  MLYCPPNVTFSQIWINNGISHCFMDTIGNGVASGFMLLMGTVQLLM...   \n",
       "3       A0A060X8Q2  MVVVQSYCEASASISEAWLEGGISPCFYFTLVPTVLLTLSFFFGAF...   \n",
       "4       A0A060XRG5  MAENMEGKVSNGEAHLDEPVANGLRHKKCSNETTVEKIKRLVMANL...   \n",
       "...            ...                                                ...   \n",
       "520258      Z9JMX9  MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...   \n",
       "520259      Z9JTY6  MTTQETRKRSTAVVRPDEERPLGDIGARVRAGVAAWTRSPALDFYA...   \n",
       "520260      Z9JU65  MNPDTITLSAGSLTTVAVIVVIALIAVAMALRFRAEVLRAATGTPA...   \n",
       "520261      Z9JXE3  MAELTIRPEDIRHALDTAVSTYQAGSTDREEIGRVTESADGIARVE...   \n",
       "520262      Z9JZ28  MTTTAPETVTATGRVARVTGAVVDIEFPPGNIPAMYNALTTVIEAT...   \n",
       "\n",
       "                                       Catalytic activity  \\\n",
       "0       CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...   \n",
       "1       CATALYTIC ACTIVITY: Reaction=CO2(out) = CO2(in...   \n",
       "2       CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...   \n",
       "3       CATALYTIC ACTIVITY: Reaction=coproporphyrin I(...   \n",
       "4       CATALYTIC ACTIVITY: Reaction=D-serine(in) + L-...   \n",
       "...                                                   ...   \n",
       "520258  CATALYTIC ACTIVITY: Reaction=K(+)(out) + ATP +...   \n",
       "520259  CATALYTIC ACTIVITY: Reaction=[GlcNAc-(1->4)-Mu...   \n",
       "520260  CATALYTIC ACTIVITY: Reaction=diphosphate + H2O...   \n",
       "520261  CATALYTIC ACTIVITY: Reaction=ATP + H2O + 4 H(+...   \n",
       "520262  CATALYTIC ACTIVITY: Reaction=ATP + H2O + 4 H(+...   \n",
       "\n",
       "                       Gene Ontology (molecular function)  Annotation  \n",
       "0       ABC-type transporter activity [GO:0140359]; AT...         5.0  \n",
       "1       ephrin receptor binding [GO:0046875]; potassiu...         5.0  \n",
       "2       ABC-type heme transporter activity [GO:0015439...         5.0  \n",
       "3       ABC-type heme transporter activity [GO:0015439...         5.0  \n",
       "4       antiporter activity [GO:0015297]; L-aspartate ...         5.0  \n",
       "...                                                   ...         ...  \n",
       "520258  ATP binding [GO:0005524]; ATP hydrolysis activ...         3.0  \n",
       "520259  glycosyltransferase activity [GO:0016757]; lip...         3.0  \n",
       "520260  diphosphate hydrolysis-driven proton transmemb...         3.0  \n",
       "520261  ADP binding [GO:0043531]; ATP binding [GO:0005...         3.0  \n",
       "520262  ATP binding [GO:0005524]; ATP hydrolysis activ...         3.0  \n",
       "\n",
       "[405820 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# (2)\n",
    "df_uniprot = df_uniprot[df_uniprot['Annotation'] >= 3.0]\n",
    "# (3)\n",
    "df_uniprot = df_uniprot.drop_duplicates(subset='Sequence', keep='first')\n",
    "# (4)\n",
    "standard_amino_acids = \"ACDEFGHIKLMNPQRSTVWY\" # 標準アミノ酸のパターンを定義\n",
    "pattern = f\"^[{standard_amino_acids}]+$\"\n",
    "df_uniprot = df_uniprot[df_uniprot['Sequence'].str.match(pattern)] # 標準アミノ酸のみを含むシーケンスを抽出\n",
    "# df_uniprotを表示\n",
    "df_uniprot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **4. Rheaデータの処理**\n",
    "\n",
    "### **(1) UniProt データから Rhea ID を抽出**\n",
    "\n",
    "-   UniProtデータから、\"Catalytic activity\"列の情報を基に、対応するRhea IDを抽出します。\n",
    "-   \"(out)\"と\"(in)\"を含む反応のみを対象とします。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>RHEA ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>coproporphyrin I(in) + ATP + H2O = coproporphy...</td>\n",
       "      <td>RHEA:66768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>coproporphyrin III(in) + ATP + H2O = coproporp...</td>\n",
       "      <td>RHEA:66664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>coproporphyrinogen III(in) + ATP + H2O = copro...</td>\n",
       "      <td>RHEA:66680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>heme b(in) + ATP + H2O = heme b(out) + ADP + p...</td>\n",
       "      <td>RHEA:19261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>pheophorbide a(in) + ATP + H2O = pheophorbide ...</td>\n",
       "      <td>RHEA:61360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651624</th>\n",
       "      <td>Z9JLF9</td>\n",
       "      <td>a quinone + NADH + 5 H(+)(in) = a quinol + NAD...</td>\n",
       "      <td>RHEA:57888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651625</th>\n",
       "      <td>Z9JMX9</td>\n",
       "      <td>K(+)(out) + ATP + H2O = K(+)(in) + ADP + phosp...</td>\n",
       "      <td>RHEA:16777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651626</th>\n",
       "      <td>Z9JU65</td>\n",
       "      <td>diphosphate + H2O + H(+)(in) = 2 phosphate + 2...</td>\n",
       "      <td>RHEA:13973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651627</th>\n",
       "      <td>Z9JXE3</td>\n",
       "      <td>ATP + H2O + 4 H(+)(in) = ADP + phosphate + 5 H...</td>\n",
       "      <td>RHEA:57720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651628</th>\n",
       "      <td>Z9JZ28</td>\n",
       "      <td>ATP + H2O + 4 H(+)(in) = ADP + phosphate + 5 H...</td>\n",
       "      <td>RHEA:57720</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>651629 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID                                           Reaction  \\\n",
       "0       A0A015JYU2  coproporphyrin I(in) + ATP + H2O = coproporphy...   \n",
       "1       A0A015JYU2  coproporphyrin III(in) + ATP + H2O = coproporp...   \n",
       "2       A0A015JYU2  coproporphyrinogen III(in) + ATP + H2O = copro...   \n",
       "3       A0A015JYU2  heme b(in) + ATP + H2O = heme b(out) + ADP + p...   \n",
       "4       A0A015JYU2  pheophorbide a(in) + ATP + H2O = pheophorbide ...   \n",
       "...            ...                                                ...   \n",
       "651624      Z9JLF9  a quinone + NADH + 5 H(+)(in) = a quinol + NAD...   \n",
       "651625      Z9JMX9  K(+)(out) + ATP + H2O = K(+)(in) + ADP + phosp...   \n",
       "651626      Z9JU65  diphosphate + H2O + H(+)(in) = 2 phosphate + 2...   \n",
       "651627      Z9JXE3  ATP + H2O + 4 H(+)(in) = ADP + phosphate + 5 H...   \n",
       "651628      Z9JZ28  ATP + H2O + 4 H(+)(in) = ADP + phosphate + 5 H...   \n",
       "\n",
       "           RHEA ID  \n",
       "0       RHEA:66768  \n",
       "1       RHEA:66664  \n",
       "2       RHEA:66680  \n",
       "3       RHEA:19261  \n",
       "4       RHEA:61360  \n",
       "...            ...  \n",
       "651624  RHEA:57888  \n",
       "651625  RHEA:16777  \n",
       "651626  RHEA:13973  \n",
       "651627  RHEA:57720  \n",
       "651628  RHEA:57720  \n",
       "\n",
       "[651629 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_rhea = \"Xref=Rhea:\"\n",
    "\n",
    "def get_reaction_info(activity):\n",
    "    # 最初のセミコロンまでの文字列を取得(反応名を取得)\n",
    "    reaction = activity.split(\";\")[0]\n",
    "    # \"Reaction=\"を取り除く\n",
    "    reaction = reaction.replace(\"Reaction=\", \"\")\n",
    "\n",
    "    # \"Xref=Rhea:\"の位置を探す\n",
    "    start_pos = activity.find(start_rhea)\n",
    "    if start_pos != -1:\n",
    "        # Rhea IDを取得\n",
    "        rhea_ID = activity[start_pos + len(start_rhea):]\n",
    "        # 最初のカンマまでを取得\n",
    "        rhea_ID = rhea_ID.split(\",\")[0]\n",
    "    else:\n",
    "        # 見つからない場合はNaNを返す\n",
    "        rhea_ID = np.nan\n",
    "\n",
    "    return(rhea_ID, reaction)\n",
    "\n",
    "def get_activtities(activities):\n",
    "    # \"CATALYTIC ACTIVITY: \"で分割\n",
    "    activities = activities.split(\"CATALYTIC ACTIVITY: \")\n",
    "    # 2番目以降の要素を取得\n",
    "    return(activities[1:])\n",
    "\n",
    "def process_row(row):\n",
    "    # \"Catalytic activity\"列を取得\n",
    "    activities = row[\"Catalytic activity\"]\n",
    "    # Nan出ないことを確認\n",
    "    if pd.notna(activities):\n",
    "        # get_activtities関数を呼び出し\n",
    "        activities = get_activtities(activities=activities)\n",
    "\n",
    "        # \"(out)\"と\"(in)\"を含む反応のみを処理\n",
    "        return [\n",
    "            (row[\"Entry\"], reaction, rhea_ID)\n",
    "            for activity in activities\n",
    "            if \"(out)\" in activity and \"(in)\" in activity\n",
    "            for rhea_ID, reaction in [get_reaction_info(activity)]\n",
    "        ]\n",
    "    return []\n",
    "\n",
    "# 各行にprocess_row関数を適用し、結果をDataFrameに変換\n",
    "df_reactions = pd.DataFrame(\n",
    "    df_uniprot.apply(process_row, axis=1).explode().dropna().tolist(),\n",
    "    columns=[\"Uniprot ID\", \"Reaction\", \"RHEA ID\"]\n",
    ")\n",
    "\n",
    "# Number of different UniProt IDs : 356595\n",
    "# Number of different RHEA reaction IDs: 1169\n",
    "\n",
    "# df_reactionsを表示\n",
    "df_reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **(2) Rheaデータから反応情報(CHEBI ID)を抽出**\n",
    "\n",
    "-   Rheaデータ（`rhea-reactions.txt`）を読み込み、各Rhea IDに対応するChEBI IDのリストを抽出します。\n",
    "-   `(out)` または `(in)` を含む基質のみを対象とします。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RHEA ID</th>\n",
       "      <th>CHEBI_ID_list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10000.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10001.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10002.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10003.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10004.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69683</th>\n",
       "      <td>82786.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69684</th>\n",
       "      <td>82787.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69685</th>\n",
       "      <td>82788.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69686</th>\n",
       "      <td>82789.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69687</th>\n",
       "      <td>82790.0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69688 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       RHEA ID CHEBI_ID_list\n",
       "0      10000.0            []\n",
       "1      10001.0            []\n",
       "2      10002.0            []\n",
       "3      10003.0            []\n",
       "4      10004.0            []\n",
       "...        ...           ...\n",
       "69683  82786.0            []\n",
       "69684  82787.0            []\n",
       "69685  82788.0            []\n",
       "69686  82789.0            []\n",
       "69687  82790.0            []\n",
       "\n",
       "[69688 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_RHEA_ID_and_CHEBI_IDs(entry):\n",
    "    # ENTRYからRHEA IDを抽出\n",
    "    RHEA_ID = entry[0][len(\"ENTRY\"): -1]\n",
    "    RHEA_ID = RHEA_ID.split(\" \")[-1]\n",
    "    # EQUATIONからCHEBI IDを抽出\n",
    "    CHEBI_IDs = entry[2][len(\"EQUATION\"): -1]\n",
    "    CHEBI_IDs = CHEBI_IDs[CHEBI_IDs.index(\"CHEBI\"):]\n",
    "\n",
    "    # DEFINITIONから基質名を抽出\n",
    "    sub_names = entry[1][len(\"DEFINITION\"): -1]\n",
    "    # 先頭に空白があれば削除\n",
    "    while sub_names[0] == \" \":\n",
    "        sub_names = sub_names[1:]\n",
    "\n",
    "    return(RHEA_ID, CHEBI_IDs, sub_names)\n",
    "\n",
    "def get_substrate_IDs(IDs):\n",
    "     # \"=\"、\"=>\"、または\"<=>\"で分割し、最初の部分を取得\n",
    "    IDs = IDs.split(\" = \")[0]\n",
    "    IDs = IDs.split(\" => \")[0]\n",
    "    IDs = IDs.split(\" <=> \")[0]\n",
    "    # \"+\"で分割し、\";\"で結合\n",
    "    IDs = IDs.replace(\" + \", \";\")\n",
    "    # \";\"で分割\n",
    "    IDs = IDs.split(\";\")\n",
    "    # 各IDの最後の部分（CHEBI ID）を取得\n",
    "    return([ID.split(\" \")[-1] for ID in IDs])\n",
    "\n",
    "def get_substrate_names(names):\n",
    "    # \"=\"、\"=>\"、または\"<=>\"で分割し、最初の部分を取得\n",
    "    names = names.split(\" = \")[0]\n",
    "    names = names.split(\" => \")[0]\n",
    "    names = names.split(\" <=> \")[0]\n",
    "    # \"+\"で分割し、\";\"で結合\n",
    "    names = names.replace(\" + \", \";\")\n",
    "    # \";\"で分割\n",
    "    names = names.split(\";\")\n",
    "     # 各IDの最後の部分（基質名）を取得\n",
    "    return([ID.split(\" \")[-1] for ID in names])\n",
    "\n",
    "# データを収集するためのリスト\n",
    "data_list = []\n",
    "\n",
    "# ファイルをストリーミング形式で読み込み\n",
    "with open('dataset/rhea-reactions.txt', 'r') as file1:\n",
    "    entry_lines = []  # 1つのエントリを蓄積するリスト\n",
    "    for line in file1:\n",
    "        if line == '///\\n':  # エントリの終わり\n",
    "            if entry_lines:  # エントリが空でない場合\n",
    "                # エントリを処理\n",
    "                RHEA_ID, CHEBI_IDs, sub_names = extract_RHEA_ID_and_CHEBI_IDs(entry_lines)\n",
    "                CHEBI_ID_list = get_substrate_IDs(IDs=CHEBI_IDs)\n",
    "                sub_name_list = get_substrate_names(names=sub_names)\n",
    "\n",
    "                # フィルタリング用のブール配列を作成\n",
    "                access_array = [(\"(out)\" in sub or \"(in)\" in sub) for sub in sub_name_list]\n",
    "                # リスト内包表記でCHEBI_ID_listをフィルタリング\n",
    "                CHEBI_ID_list = [chebi for chebi, access in zip(CHEBI_ID_list, access_array) if access]\n",
    "\n",
    "                # RHEA IDを即座に数値に変換\n",
    "                RHEA_ID_num = float(RHEA_ID.split(\":\")[-1])\n",
    "\n",
    "                # データをリストに追加\n",
    "                data_list.append({\"RHEA ID\": RHEA_ID_num, \"CHEBI_ID_list\": CHEBI_ID_list})\n",
    "\n",
    "                entry_lines = []  # エントリをリセット\n",
    "        else:\n",
    "            entry_lines.append(line)  # エントリの行を蓄積\n",
    "\n",
    "# データをデータフレームに変換\n",
    "df_RHEA = pd.DataFrame(data_list)\n",
    "# df_RHEAを表示\n",
    "df_RHEA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **(3) Rheaデータを統合**\n",
    "\n",
    "- `df_reactions`と`df_RHEA`を`RHEA ID`で結合\n",
    "- 計算効率向上のため`lru_cache`を使用"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 651629/651629 [00:19<00:00, 33728.90it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>ChEBI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:167478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:131725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:57309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:60344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:58687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788531</th>\n",
       "      <td>Z9JLF9</td>\n",
       "      <td>CHEBI:15378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788532</th>\n",
       "      <td>Z9JMX9</td>\n",
       "      <td>CHEBI:29103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788533</th>\n",
       "      <td>Z9JU65</td>\n",
       "      <td>CHEBI:15378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788534</th>\n",
       "      <td>Z9JXE3</td>\n",
       "      <td>CHEBI:15378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788535</th>\n",
       "      <td>Z9JZ28</td>\n",
       "      <td>CHEBI:15378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>646734 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID         ChEBI\n",
       "0       A0A015JYU2  CHEBI:167478\n",
       "1       A0A015JYU2  CHEBI:131725\n",
       "2       A0A015JYU2   CHEBI:57309\n",
       "3       A0A015JYU2   CHEBI:60344\n",
       "4       A0A015JYU2   CHEBI:58687\n",
       "...            ...           ...\n",
       "788531      Z9JLF9   CHEBI:15378\n",
       "788532      Z9JMX9   CHEBI:29103\n",
       "788533      Z9JU65   CHEBI:15378\n",
       "788534      Z9JXE3   CHEBI:15378\n",
       "788535      Z9JZ28   CHEBI:15378\n",
       "\n",
       "[646734 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from functools import lru_cache\n",
    "from tqdm import tqdm\n",
    "\n",
    "def get_CHEBI_IDs_from_RHEA_ID(rhea_ID):\n",
    "    rhea_ID = float(rhea_ID.replace(\"RHEA:\", \"\"))\n",
    "    try:\n",
    "        CHEBI_IDs = list(df_RHEA[\"CHEBI_ID_list\"].loc[df_RHEA[\"RHEA ID\"] == rhea_ID])[0]\n",
    "    except IndexError:\n",
    "        CHEBI_IDs = []\n",
    "    return(CHEBI_IDs)\n",
    "\n",
    "# 関数をキャッシュする\n",
    "@lru_cache(maxsize=None)\n",
    "def cached_get_CHEBI_IDs(rhea_id):\n",
    "    return get_CHEBI_IDs_from_RHEA_ID(rhea_ID=rhea_id)\n",
    "\n",
    "# より効率的なループバージョン\n",
    "def optimized_loop():\n",
    "    uniprot_ids = []\n",
    "    chebi_ids = []\n",
    "    \n",
    "    # NaN値でないRHEA IDを持つ行だけを処理\n",
    "    filtered_df = df_reactions.dropna(subset=[\"RHEA ID\"])\n",
    "    \n",
    "    for _, row in tqdm(filtered_df.iterrows(), total=len(filtered_df)):\n",
    "        rhea_id = row[\"RHEA ID\"]\n",
    "        uid = row[\"Uniprot ID\"]\n",
    "        \n",
    "        chebi_id_list = cached_get_CHEBI_IDs(rhea_id)\n",
    "        \n",
    "        # 各CHEBI IDに対してリストに追加\n",
    "        for chebi_id in chebi_id_list:\n",
    "            uniprot_ids.append(uid)\n",
    "            chebi_ids.append(chebi_id)\n",
    "    \n",
    "    # 一度にデータフレームを作成\n",
    "    df_prot_sub = pd.DataFrame({\"Uniprot ID\": uniprot_ids, \"ChEBI\": chebi_ids})\n",
    "    df_prot_sub.drop_duplicates(inplace=True)\n",
    "    return df_prot_sub\n",
    "\n",
    "# 関数を実行\n",
    "df_prot_sub = optimized_loop()\n",
    "# df_prot_subを表示\n",
    "df_prot_sub"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **5. ChEBIデータの処理**\n",
    "\n",
    "### **(1) ChEBI IDと基質名の対応表を読み込み**\n",
    "\n",
    "-   ChEBI IDと基質名の対応表（`compounds.tsv`）を読み込みます。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ChEBI</th>\n",
       "      <th>Substrate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEBI:9349</td>\n",
       "      <td>sulfonyldimethane</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEBI:9352</td>\n",
       "      <td>sulindac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEBI:9355</td>\n",
       "      <td>sulfuretin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEBI:9380</td>\n",
       "      <td>syringin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEBI:9427</td>\n",
       "      <td>2-methylanthraquinone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228011</th>\n",
       "      <td>CHEBI:707880</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228012</th>\n",
       "      <td>CHEBI:718203</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228013</th>\n",
       "      <td>CHEBI:741566</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228014</th>\n",
       "      <td>CHEBI:232829</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228015</th>\n",
       "      <td>CHEBI:232578</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>228016 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               ChEBI              Substrate\n",
       "0         CHEBI:9349      sulfonyldimethane\n",
       "1         CHEBI:9352               sulindac\n",
       "2         CHEBI:9355             sulfuretin\n",
       "3         CHEBI:9380               syringin\n",
       "4         CHEBI:9427  2-methylanthraquinone\n",
       "...              ...                    ...\n",
       "228011  CHEBI:707880                    NaN\n",
       "228012  CHEBI:718203                    NaN\n",
       "228013  CHEBI:741566                    NaN\n",
       "228014  CHEBI:232829                    NaN\n",
       "228015  CHEBI:232578                    NaN\n",
       "\n",
       "[228016 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compounds.tsvを読み込み\n",
    "df_substrate = pd.read_csv('dataset/compounds.tsv', sep='\\t', encoding='latin-1')\n",
    "# 必要な列を抽出\n",
    "df_substrate = df_substrate[[\"CHEBI_ACCESSION\", \"NAME\"]]\n",
    "# 列名を変更\n",
    "df_substrate.columns = [\"ChEBI\", \"Substrate\"]\n",
    "# df_substrateを表示\n",
    "df_substrate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **(2) UniProtデータとChEBI IDを統合**\n",
    "\n",
    "-   `df_prot_sub`と`df_substrate`をChEBI IDをキーにして統合し、`df_transporter`を作成します。\n",
    "-   `df_transporter`には、UniProt ID、ChEBI ID、基質名（Substrate）が含まれます。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>ChEBI</th>\n",
       "      <th>Substrate</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:167478</td>\n",
       "      <td>coproporphyrin I(4-)</td>\n",
       "      <td>MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:131725</td>\n",
       "      <td>coproporphyrin III(4-)</td>\n",
       "      <td>MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:57309</td>\n",
       "      <td>coproporphyrinogen III(4-)</td>\n",
       "      <td>MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:60344</td>\n",
       "      <td>ferroheme b(2-)</td>\n",
       "      <td>MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A015JYU2</td>\n",
       "      <td>CHEBI:58687</td>\n",
       "      <td>pheophorbide a(2-)</td>\n",
       "      <td>MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646729</th>\n",
       "      <td>Z9JLF9</td>\n",
       "      <td>CHEBI:15378</td>\n",
       "      <td>hydron</td>\n",
       "      <td>MLAEYLPSLLFLVVATGIGLALILIGRFLGPHRPDFEKLSSYECGF...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646730</th>\n",
       "      <td>Z9JMX9</td>\n",
       "      <td>CHEBI:29103</td>\n",
       "      <td>potassium(1+)</td>\n",
       "      <td>MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646731</th>\n",
       "      <td>Z9JU65</td>\n",
       "      <td>CHEBI:15378</td>\n",
       "      <td>hydron</td>\n",
       "      <td>MNPDTITLSAGSLTTVAVIVVIALIAVAMALRFRAEVLRAATGTPA...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646732</th>\n",
       "      <td>Z9JXE3</td>\n",
       "      <td>CHEBI:15378</td>\n",
       "      <td>hydron</td>\n",
       "      <td>MAELTIRPEDIRHALDTAVSTYQAGSTDREEIGRVTESADGIARVE...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646733</th>\n",
       "      <td>Z9JZ28</td>\n",
       "      <td>CHEBI:15378</td>\n",
       "      <td>hydron</td>\n",
       "      <td>MTTTAPETVTATGRVARVTGAVVDIEFPPGNIPAMYNALTTVIEAT...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>646734 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID         ChEBI                   Substrate  \\\n",
       "0       A0A015JYU2  CHEBI:167478        coproporphyrin I(4-)   \n",
       "1       A0A015JYU2  CHEBI:131725      coproporphyrin III(4-)   \n",
       "2       A0A015JYU2   CHEBI:57309  coproporphyrinogen III(4-)   \n",
       "3       A0A015JYU2   CHEBI:60344             ferroheme b(2-)   \n",
       "4       A0A015JYU2   CHEBI:58687          pheophorbide a(2-)   \n",
       "...            ...           ...                         ...   \n",
       "646729      Z9JLF9   CHEBI:15378                      hydron   \n",
       "646730      Z9JMX9   CHEBI:29103               potassium(1+)   \n",
       "646731      Z9JU65   CHEBI:15378                      hydron   \n",
       "646732      Z9JXE3   CHEBI:15378                      hydron   \n",
       "646733      Z9JZ28   CHEBI:15378                      hydron   \n",
       "\n",
       "                                                 Sequence  Annotation  \n",
       "0       MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...         5.0  \n",
       "1       MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...         5.0  \n",
       "2       MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...         5.0  \n",
       "3       MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...         5.0  \n",
       "4       MKNLPVSFSAESPGKVLFSPTSGISLFSLFPLFLFVPVFLLSFIIL...         5.0  \n",
       "...                                                   ...         ...  \n",
       "646729  MLAEYLPSLLFLVVATGIGLALILIGRFLGPHRPDFEKLSSYECGF...         3.0  \n",
       "646730  MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...         3.0  \n",
       "646731  MNPDTITLSAGSLTTVAVIVVIALIAVAMALRFRAEVLRAATGTPA...         3.0  \n",
       "646732  MAELTIRPEDIRHALDTAVSTYQAGSTDREEIGRVTESADGIARVE...         3.0  \n",
       "646733  MTTTAPETVTATGRVARVTGAVVDIEFPPGNIPAMYNALTTVIEAT...         3.0  \n",
       "\n",
       "[646734 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df_prot_subとdf_substrateを\"ChEBI\"列でマージ\n",
    "df_transporter = df_prot_sub.merge(df_substrate[[\"ChEBI\", \"Substrate\"]], on=\"ChEBI\", how=\"left\")\n",
    "# さらにdf_uniprotと\"Uniprot ID\"でマージ\n",
    "df_transporter = df_transporter.merge(df_uniprot[[\"Entry\", \"Sequence\", \"Annotation\"]], left_on=\"Uniprot ID\", right_on=\"Entry\", how=\"left\")\n",
    "# 不要な列を削除\n",
    "df_transporter = df_transporter.drop(columns=[\"Entry\"])\n",
    "# df_transporterを表示\n",
    "df_transporter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **(3) substrate_countsの確認**\n",
    "\n",
    "-   基質の種類ごとの出現回数をカウントし、ファイルに保存します。\n",
    "-   これにより、データセット内の基質の分布を確認できます。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_transporter['Substrate'].value_counts()を保存\n",
    "df_transporter['Substrate'].value_counts().to_csv('dataset/substrate_counts.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **6. 最終データセットの作成**\n",
    "\n",
    "### **(1) 1行1配列に変換**\n",
    "\n",
    "-   `df_transporter`を、UniProt IDと配列（Sequence）ごとに1行になるように変換します。\n",
    "-   各行には、UniProt ID、配列、対応する基質（複数存在する場合はリスト形式）が含まれます。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Substrates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009HCH1</td>\n",
       "      <td>MATIHVDGKSYEVNGSENLLQACLSLGIDIPYFCWHPSLGSVGSCR...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[hydron]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009HCH6</td>\n",
       "      <td>MSAITPYDWAIIAFVIGVTFLCVFMLTVPLLLGGKSWGRAKQEQFE...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[hydron]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A009HEB6</td>\n",
       "      <td>MSSGRIIQIIGAVIDVEFERNSVPKIYDALQVDGTETTLEVQQQLG...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[hydron]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A009HI66</td>\n",
       "      <td>MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKNPVMAIVWLG...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A009HL19</td>\n",
       "      <td>MSSGRIIQIIGAVIDVEFERTSVPKIYDALQVDGTETTLEVQQQLG...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[hydron]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356590</th>\n",
       "      <td>Z9JLF9</td>\n",
       "      <td>MLAEYLPSLLFLVVATGIGLALILIGRFLGPHRPDFEKLSSYECGF...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[hydron]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356591</th>\n",
       "      <td>Z9JMX9</td>\n",
       "      <td>MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356592</th>\n",
       "      <td>Z9JU65</td>\n",
       "      <td>MNPDTITLSAGSLTTVAVIVVIALIAVAMALRFRAEVLRAATGTPA...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[hydron]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356593</th>\n",
       "      <td>Z9JXE3</td>\n",
       "      <td>MAELTIRPEDIRHALDTAVSTYQAGSTDREEIGRVTESADGIARVE...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[hydron]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356594</th>\n",
       "      <td>Z9JZ28</td>\n",
       "      <td>MTTTAPETVTATGRVARVTGAVVDIEFPPGNIPAMYNALTTVIEAT...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[hydron]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>356595 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID                                           Sequence  \\\n",
       "0       A0A009HCH1  MATIHVDGKSYEVNGSENLLQACLSLGIDIPYFCWHPSLGSVGSCR...   \n",
       "1       A0A009HCH6  MSAITPYDWAIIAFVIGVTFLCVFMLTVPLLLGGKSWGRAKQEQFE...   \n",
       "2       A0A009HEB6  MSSGRIIQIIGAVIDVEFERNSVPKIYDALQVDGTETTLEVQQQLG...   \n",
       "3       A0A009HI66  MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKNPVMAIVWLG...   \n",
       "4       A0A009HL19  MSSGRIIQIIGAVIDVEFERTSVPKIYDALQVDGTETTLEVQQQLG...   \n",
       "...            ...                                                ...   \n",
       "356590      Z9JLF9  MLAEYLPSLLFLVVATGIGLALILIGRFLGPHRPDFEKLSSYECGF...   \n",
       "356591      Z9JMX9  MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...   \n",
       "356592      Z9JU65  MNPDTITLSAGSLTTVAVIVVIALIAVAMALRFRAEVLRAATGTPA...   \n",
       "356593      Z9JXE3  MAELTIRPEDIRHALDTAVSTYQAGSTDREEIGRVTESADGIARVE...   \n",
       "356594      Z9JZ28  MTTTAPETVTATGRVARVTGAVVDIEFPPGNIPAMYNALTTVIEAT...   \n",
       "\n",
       "        Annotation       Substrates  \n",
       "0              3.0         [hydron]  \n",
       "1              3.0         [hydron]  \n",
       "2              3.0         [hydron]  \n",
       "3              3.0  [potassium(1+)]  \n",
       "4              3.0         [hydron]  \n",
       "...            ...              ...  \n",
       "356590         3.0         [hydron]  \n",
       "356591         3.0  [potassium(1+)]  \n",
       "356592         3.0         [hydron]  \n",
       "356593         3.0         [hydron]  \n",
       "356594         3.0         [hydron]  \n",
       "\n",
       "[356595 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df_transporterのUniprot ID, Sequence, Substrateを取得\n",
    "df_transporter2 = df_transporter[[\"Uniprot ID\", \"Sequence\", \"Substrate\", \"Annotation\"]]\n",
    "# Uniprot IDとSequenceごとにグループ化し、Substrateをリスト化\n",
    "df_unique = df_transporter2.groupby(['Uniprot ID', 'Sequence', \"Annotation\"], as_index=False).agg({'Substrate': list})\n",
    "# Substrateカラム名をSubstratesに変更\n",
    "df_unique = df_unique.rename(columns={'Substrate': 'Substrates'})\n",
    "# df_uniqueを表示\n",
    "df_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **(2) hydronを削除**\n",
    "\n",
    "-   `hydron`のみを基質とするエントリが多すぎるため、`Annotation`が3.0のエントリを削除します。\n",
    "-   これにより、データセットのバランスを調整します。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Substrates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A009HI66</td>\n",
       "      <td>MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKNPVMAIVWLG...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A0A009HS66</td>\n",
       "      <td>MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKKPVMAIVWLG...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A0A009ITI1</td>\n",
       "      <td>MKGNRDVINQLNQVLYHHLTAINQYFLHSRMFNDWGIEQLGSAEYK...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[iron(2+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A0A009MN79</td>\n",
       "      <td>MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKNPVMAIVWLG...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A0A009PGQ3</td>\n",
       "      <td>MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKNPVMAIVWLG...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356573</th>\n",
       "      <td>X8JSI9</td>\n",
       "      <td>MTGTVGSSFSEATLGPLKSTLVKDLHISNARYGTISSATSVINMFL...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356576</th>\n",
       "      <td>Y0KH45</td>\n",
       "      <td>MSSPAIKSLLDPTLVKPAIVDSFKKLAPQIQWRNPVMFVVYLGSLL...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356582</th>\n",
       "      <td>Z4WWK6</td>\n",
       "      <td>MEQLRSSLRDKPAARWGALALVSLTMFFAYMFVDVLSPVKILVEKE...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356583</th>\n",
       "      <td>Z4YJD4</td>\n",
       "      <td>MEKWASLDSDESEPPAQENSCPDPPDRDPNSKPPPAKPHIFATRSR...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[calcium(2+), potassium(1+), magnesium(2+), so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356591</th>\n",
       "      <td>Z9JMX9</td>\n",
       "      <td>MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106834 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID                                           Sequence  \\\n",
       "3       A0A009HI66  MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKNPVMAIVWLG...   \n",
       "7       A0A009HS66  MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKKPVMAIVWLG...   \n",
       "18      A0A009ITI1  MKGNRDVINQLNQVLYHHLTAINQYFLHSRMFNDWGIEQLGSAEYK...   \n",
       "20      A0A009MN79  MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKNPVMAIVWLG...   \n",
       "22      A0A009PGQ3  MNTQTHVNSHKQTMAVPNFEVWKNAFVKLLPQHAIKNPVMAIVWLG...   \n",
       "...            ...                                                ...   \n",
       "356573      X8JSI9  MTGTVGSSFSEATLGPLKSTLVKDLHISNARYGTISSATSVINMFL...   \n",
       "356576      Y0KH45  MSSPAIKSLLDPTLVKPAIVDSFKKLAPQIQWRNPVMFVVYLGSLL...   \n",
       "356582      Z4WWK6  MEQLRSSLRDKPAARWGALALVSLTMFFAYMFVDVLSPVKILVEKE...   \n",
       "356583      Z4YJD4  MEKWASLDSDESEPPAQENSCPDPPDRDPNSKPPPAKPHIFATRSR...   \n",
       "356591      Z9JMX9  MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...   \n",
       "\n",
       "        Annotation                                         Substrates  \n",
       "3              3.0                                    [potassium(1+)]  \n",
       "7              3.0                                    [potassium(1+)]  \n",
       "18             3.0                                         [iron(2+)]  \n",
       "20             3.0                                    [potassium(1+)]  \n",
       "22             3.0                                    [potassium(1+)]  \n",
       "...            ...                                                ...  \n",
       "356573         4.0  [Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...  \n",
       "356576         3.0                                    [potassium(1+)]  \n",
       "356582         4.0  [Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...  \n",
       "356583         3.0  [calcium(2+), potassium(1+), magnesium(2+), so...  \n",
       "356591         3.0                                    [potassium(1+)]  \n",
       "\n",
       "[106834 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 削除したい条件に合致する行のインデックスを取得\n",
    "mask = (df_unique['Substrates'].apply(lambda x: x == ['hydron'])) & (df_unique['Annotation'] == 3.0)\n",
    "indices_to_cut = df_unique[mask].index\n",
    "\n",
    "# 選ばれたインデックスを除外\n",
    "df_reduced = df_unique.drop(indices_to_cut)\n",
    "# df_reducedを表示\n",
    "df_reduced"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **(3) substrate_countsの確認**\n",
    "\n",
    "-   `hydron`削除後の基質の数をカウントし、基質ごとのエントリ数を確認します。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# substrates_counts_after_reducing_hydron.csvとして保存\n",
    "df_reduced['Substrates'].value_counts().to_csv('dataset/substrates_counts_after_reducing_hydron.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "\n",
    "### **(追加) 90%以上の類似性でクラスタリングし代表配列のみ抽出**\n",
    "\n",
    "-  配列をFASTA形式で保存し、MMseqs2で90%以上の類似性でクラスタリングを実行します。\n",
    "-  MMseqs2のクラスタリング結果から代表配列のみを抽出します。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster dataset/uniprot_sequences.fasta dataset/clustered_uniprot tmp --min-seq-id 0.9 \n",
      "\n",
      "MMseqs Version:                     \t13.45111\n",
      "Substitution matrix                 \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Seed substitution matrix            \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "k-score                             \t2147483647\n",
      "Alphabet size                       \tnucl:5,aa:21\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \t1\n",
      "Pseudo count b                      \t1.5\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t36\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.9\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Gap open cost                       \tnucl:5,aa:11\n",
      "Gap extension cost                  \tnucl:2,aa:1\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \tnucl:0.200,aa:0.000\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb dataset/uniprot_sequences.fasta tmp/17472568689772865097/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Shuffle database cannot be combined with --createdb-mode 0\n",
      "We recompute with --shuffle 0\n",
      "Converting sequences\n",
      "[==========\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 57ms\n",
      "Create directory tmp/17472568689772865097/clu_tmp\n",
      "cluster tmp/17472568689772865097/input tmp/17472568689772865097/clu tmp/17472568689772865097/clu_tmp --max-seqs 20 -c 0.8 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.9 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust tmp/17472568689772865097/input tmp/17472568689772865097/clu_tmp/17492949728288354519/clu_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 36 --compressed 0 -v 3 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --alph-size nucl:5,aa:13 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher tmp/17472568689772865097/input tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "kmermatcher tmp/17472568689772865097/input tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Database size: 106834 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 106.83K 0s 120ms\n",
      "Sort kmer 0h 0m 0s 36ms\n",
      "Sort by rep. sequence 0h 0m 0s 20ms\n",
      "Time for fill: 0h 0m 0s 32ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 269ms\n",
      "rescorediagonal tmp/17472568689772865097/input tmp/17472568689772865097/input tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_rescore1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 106.83K 0s 42ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 39ms\n",
      "Time for processing: 0h 0m 0s 220ms\n",
      "clust tmp/17472568689772865097/input tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_rescore1 tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 106.83K 0s 6ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 85030 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 106.83K 0s 6ms\n",
      "Add missing connections\n",
      "[=================================================================] 106.83K 0s 5ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 35ms\n",
      "Total time: 0h 0m 0s 65ms\n",
      "\n",
      "Size of the sequence database: 106834\n",
      "Size of the alignment database: 106834\n",
      "Number of clusters: 61024\n",
      "\n",
      "Writing results 0h 0m 0s 8ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 95ms\n",
      "createsubdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/order_redundancy tmp/17472568689772865097/input tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "createsubdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/order_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 31ms\n",
      "filterdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_filter1 tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_filter2 --filter-file tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/order_redundancy --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 61.02K 0s 20ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 14ms\n",
      "Time for processing: 0h 0m 0s 142ms\n",
      "rescorediagonal tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_filter2 tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_rescore2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 61.02K 0s 25ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 13ms\n",
      "Time for processing: 0h 0m 0s 142ms\n",
      "align tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_rescore2 tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/aln --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 61024 type: Aminoacid\n",
      "Target database size: 61024 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 61.02K 3s 713ms\n",
      "Time for merging to aln: 0h 0m 0s 15ms\n",
      "146961 alignments calculated\n",
      "76861 sequence pairs passed the thresholds (0.523003 of overall calculated)\n",
      "1.259521 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 503ms\n",
      "clust tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/aln tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 61.02K 0s 4ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 15837 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 61.02K 0s 4ms\n",
      "Add missing connections\n",
      "[=================================================================] 61.02K 0s 2ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 21ms\n",
      "Total time: 0h 0m 0s 36ms\n",
      "\n",
      "Size of the sequence database: 61024\n",
      "Size of the alignment database: 61024\n",
      "Number of clusters: 49580\n",
      "\n",
      "Writing results 0h 0m 0s 6ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 59ms\n",
      "mergeclusters tmp/17472568689772865097/input tmp/17472568689772865097/clu_tmp/17492949728288354519/clu_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pre_clust tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/clust --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 61.02K 0s 9ms\n",
      "Clustering step 2\n",
      "[=================================================================] 49.58K 0s 20ms\n",
      "Write merged clustering\n",
      "[=================================================================] 106.83K 0s 67ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 11ms\n",
      "Time for processing: 0h 0m 0s 93ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/linclust/11246476238104388143/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb tmp/17472568689772865097/clu_tmp/17492949728288354519/clu_redundancy tmp/17472568689772865097/input tmp/17472568689772865097/clu_tmp/17492949728288354519/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "prefilter tmp/17472568689772865097/clu_tmp/17492949728288354519/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/pref_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 1 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 0 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 49580 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 49580 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 49.58K 0s 248ms\n",
      "Index table: Masked residues: 213023\n",
      "Index table: fill\n",
      "[=================================================================] 49.58K 0s 101ms\n",
      "Index statistics\n",
      "Entries:          10747521\n",
      "DB size:          549 MB\n",
      "Avg k-mer size:   0.167930\n",
      "Top 10 k-mers\n",
      "    AMGGDA\t7642\n",
      "    NDPLAD\t7572\n",
      "    FSADKY\t5186\n",
      "    ADGDGK\t4930\n",
      "    NMDDPT\t4912\n",
      "    GDFDGP\t4338\n",
      "    DPKTPS\t4207\n",
      "    LFRGIY\t3580\n",
      "    GGYKDV\t3563\n",
      "    FAVGYT\t3220\n",
      "Time for index table init: 0h 0m 0s 786ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 49580\n",
      "Target db start 1 to 49580\n",
      "[=================================================================] 49.58K 1s 592ms\n",
      "\n",
      "1.707048 k-mers per position\n",
      "43257 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "19 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 12ms\n",
      "Time for processing: 0h 0m 2s 605ms\n",
      "align tmp/17472568689772865097/clu_tmp/17492949728288354519/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/pref_step0 tmp/17472568689772865097/clu_tmp/17492949728288354519/aln_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 49580 type: Aminoacid\n",
      "Target database size: 49580 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 49.58K 30s 665ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 12ms\n",
      "861436 alignments calculated\n",
      "139901 sequence pairs passed the thresholds (0.162404 of overall calculated)\n",
      "2.821723 hits per query sequence\n",
      "Time for processing: 0h 0m 30s 768ms\n",
      "clust tmp/17472568689772865097/clu_tmp/17492949728288354519/input_step_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/aln_step0 tmp/17472568689772865097/clu_tmp/17492949728288354519/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 49.58K 0s 4ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 20589 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 49.58K 0s 4ms\n",
      "Add missing connections\n",
      "[=================================================================] 49.58K 0s 3ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 20ms\n",
      "Total time: 0h 0m 0s 32ms\n",
      "\n",
      "Size of the sequence database: 49580\n",
      "Size of the alignment database: 49580\n",
      "Number of clusters: 37998\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 60ms\n",
      "mergeclusters tmp/17472568689772865097/input tmp/17472568689772865097/clu tmp/17472568689772865097/clu_tmp/17492949728288354519/clu_redundancy tmp/17472568689772865097/clu_tmp/17492949728288354519/clu_step0 --threads 36 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 49.58K 0s 4ms\n",
      "Clustering step 2\n",
      "[=================================================================] 38.00K 0s 14ms\n",
      "Write merged clustering\n",
      "[=================================================================] 106.83K 0s 128ms\n",
      "Time for merging to clu: 0h 0m 0s 9ms\n",
      "Time for processing: 0h 0m 0s 165ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/17472568689772865097/clu_tmp/17492949728288354519/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv tmp/17472568689772865097/input tmp/17472568689772865097/input tmp/17472568689772865097/clu tmp/17472568689772865097/cluster.tsv --threads 36 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 13ms\n",
      "Time for processing: 0h 0m 0s 152ms\n",
      "result2repseq tmp/17472568689772865097/input tmp/17472568689772865097/clu tmp/17472568689772865097/clu_rep --db-load-mode 0 --compressed 0 --threads 36 -v 3 \n",
      "\n",
      "[=================================================================] 38.00K 0s 6ms\n",
      "Time for merging to clu_rep: 0h 0m 0s 24ms\n",
      "Time for processing: 0h 0m 0s 155ms\n",
      "result2flat tmp/17472568689772865097/input tmp/17472568689772865097/input tmp/17472568689772865097/clu_rep tmp/17472568689772865097/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 90ms\n",
      "createseqfiledb tmp/17472568689772865097/input tmp/17472568689772865097/clu tmp/17472568689772865097/clu_seqs --threads 36 -v 3 \n",
      "\n",
      "[=================================================================] 38.00K 0s 18ms\n",
      "Time for merging to clu_seqs: 0h 0m 0s 9ms\n",
      "Time for processing: 0h 0m 0s 177ms\n",
      "result2flat tmp/17472568689772865097/input tmp/17472568689772865097/input tmp/17472568689772865097/clu_seqs tmp/17472568689772865097/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 175ms\n",
      "rmdb tmp/17472568689772865097/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17472568689772865097/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb tmp/17472568689772865097/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb tmp/17472568689772865097/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    }
   ],
   "source": [
    "# 1. DataFrame から FASTA ファイルを生成\n",
    "fasta_file = \"dataset/uniprot_sequences.fasta\"\n",
    "with open(fasta_file, \"w\") as f:\n",
    "    for _, row in df_reduced.iterrows():\n",
    "        f.write(f\">{row['Uniprot ID']}\\n{row['Sequence']}\\n\")\n",
    "\n",
    "# 2. MMseqs2 でクラスタリングを実行\n",
    "output_prefix = \"dataset/clustered_uniprot\"\n",
    "tmp_dir = \"tmp\"\n",
    "\n",
    "subprocess.run([\n",
    "    \"mmseqs\", \"easy-cluster\", fasta_file, output_prefix, tmp_dir,\n",
    "    \"--min-seq-id\", \"0.9\"\n",
    "])\n",
    "\n",
    "# 3. クラスタリング結果を解析\n",
    "cluster_file = f\"{output_prefix}_cluster.tsv\"\n",
    "clusters = pd.read_csv(cluster_file, sep=\"\\t\", names=[\"cluster_rep\", \"member\"])\n",
    "\n",
    "\n",
    "rep_sequences = clusters[\"cluster_rep\"].unique() # 代表配列のリストを取得\n",
    "\n",
    "# 4. DataFrame をフィルタリング\n",
    "df_reduced = df_reduced[df_reduced['Uniprot ID'].isin(rep_sequences)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Substrates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>A0A009YQY2</td>\n",
       "      <td>MSTQIQSQQTSQQRIEAIKSAFVKLLPQHMLKNPVMGIVWLGTLIT...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>A0A010R4L0</td>\n",
       "      <td>MISKDPVERPLPVADGEDNASPTRSETDEKGIADATTYPSDDSESI...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>A0A010RW16</td>\n",
       "      <td>MTRKAFSLFDPKLIRPAMVESVRKLSPRIQWRNPVMFVVYIGAIIT...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>A0A010S401</td>\n",
       "      <td>MSYAFRRVCLQLPASFGPKHGPLRQAFRPAGPGIVTRQIPVIRRAF...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[calcium(2+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>A0A010S454</td>\n",
       "      <td>MPDTPESPVPVQDEEKTQPDIDIQALGISPADLPPIDTEDETLQRI...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356556</th>\n",
       "      <td>X8BF62</td>\n",
       "      <td>MQGDSDVLRLLNEQLTSELTAINQYFLHSKMQENWGFTKVAERTRA...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[iron(2+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356567</th>\n",
       "      <td>X8GY71</td>\n",
       "      <td>MELFFLIGFIGALVALLFAQSRRLVVMACPEGNDRMRKIASAIREG...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[sodium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356576</th>\n",
       "      <td>Y0KH45</td>\n",
       "      <td>MSSPAIKSLLDPTLVKPAIVDSFKKLAPQIQWRNPVMFVVYLGSLL...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356582</th>\n",
       "      <td>Z4WWK6</td>\n",
       "      <td>MEQLRSSLRDKPAARWGALALVSLTMFFAYMFVDVLSPVKILVEKE...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356591</th>\n",
       "      <td>Z9JMX9</td>\n",
       "      <td>MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[potassium(1+)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37998 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID                                           Sequence  \\\n",
       "33      A0A009YQY2  MSTQIQSQQTSQQRIEAIKSAFVKLLPQHMLKNPVMGIVWLGTLIT...   \n",
       "43      A0A010R4L0  MISKDPVERPLPVADGEDNASPTRSETDEKGIADATTYPSDDSESI...   \n",
       "45      A0A010RW16  MTRKAFSLFDPKLIRPAMVESVRKLSPRIQWRNPVMFVVYIGAIIT...   \n",
       "48      A0A010S401  MSYAFRRVCLQLPASFGPKHGPLRQAFRPAGPGIVTRQIPVIRRAF...   \n",
       "49      A0A010S454  MPDTPESPVPVQDEEKTQPDIDIQALGISPADLPPIDTEDETLQRI...   \n",
       "...            ...                                                ...   \n",
       "356556      X8BF62  MQGDSDVLRLLNEQLTSELTAINQYFLHSKMQENWGFTKVAERTRA...   \n",
       "356567      X8GY71  MELFFLIGFIGALVALLFAQSRRLVVMACPEGNDRMRKIASAIREG...   \n",
       "356576      Y0KH45  MSSPAIKSLLDPTLVKPAIVDSFKKLAPQIQWRNPVMFVVYLGSLL...   \n",
       "356582      Z4WWK6  MEQLRSSLRDKPAARWGALALVSLTMFFAYMFVDVLSPVKILVEKE...   \n",
       "356591      Z9JMX9  MAHASTKSLFTAELIVPAIGDAFKKLNPKELIRNPVMFVTACVALL...   \n",
       "\n",
       "        Annotation                                         Substrates  \n",
       "33             3.0                                    [potassium(1+)]  \n",
       "43             4.0  [Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...  \n",
       "45             3.0                                    [potassium(1+)]  \n",
       "48             3.0                                      [calcium(2+)]  \n",
       "49             4.0  [Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...  \n",
       "...            ...                                                ...  \n",
       "356556         3.0                                         [iron(2+)]  \n",
       "356567         3.0                                       [sodium(1+)]  \n",
       "356576         3.0                                    [potassium(1+)]  \n",
       "356582         4.0  [Ala-Lys(1+), L-aminoacyl-L-arginine(1+), L-am...  \n",
       "356591         3.0                                    [potassium(1+)]  \n",
       "\n",
       "[37998 rows x 4 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_reduced"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **7. 学習データセットの生成**\n",
    "\n",
    "### **(1) データの分割と整形と保存**\n",
    "\n",
    "-   データをランダムにトレーニングデータセット(90%)とテストデータセット(10％)にランダム分割し、JSON形式で保存します。\n",
    "-   保存の際、以下の形式でデータを整形します。\n",
    "\n",
    "    ```json\n",
    "    {\n",
    "        \"instruction\": \"[Generate by Substrate]\",\n",
    "        \"input\": \"Substrate=<substrate1>, <substrate2>, ...\",\n",
    "        \"output\": \"Seq=<sequence>\"\n",
    "    }\n",
    "    ```\n",
    "\n",
    "    ```json\n",
    "    {\n",
    "        \"instruction\": \"[Determine Substrate]\",\n",
    "        \"input\": \"Seq=<sequence>\",\n",
    "        \"output\": \"Substrate=<substrate1>, <substrate2>, ...\"\n",
    "    }\n",
    "    ```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "import json\n",
    "\n",
    "def save_to_json_split(sequences, substrates, test_size=0.1, random_state=0, save_dir=\".\"):\n",
    "    \"\"\"\n",
    "    データをトレーニング用とテスト用に分割し、JSONファイルに保存する関数。\n",
    "\n",
    "    Args:\n",
    "        sequences: タンパク質配列のリスト。\n",
    "        substrates: 基質名のリストのリスト。\n",
    "        test_size: テストデータの割合 (例: 0.2 なら20%)。\n",
    "        random_state: ランダムシード。\n",
    "        save_dir: JSONファイルを保存するディレクトリ。\n",
    "    \"\"\"\n",
    "\n",
    "    # データをトレーニング用とテスト用に分割\n",
    "    train_sequences, test_sequences, train_substrates, test_substrates = train_test_split(\n",
    "        sequences, substrates, test_size=test_size, random_state=random_state\n",
    "    )\n",
    "\n",
    "    # トレーニングデータを作成して保存\n",
    "    train_data_list = []\n",
    "    for i in range(len(train_sequences)):\n",
    "        substrates_string = \", \".join(train_substrates[i])\n",
    "        substrates_string_formatted = f\"Substrate=<{substrates_string}>\"\n",
    "\n",
    "        # 基質から配列を生成する指示\n",
    "        data_point_generate = {\n",
    "            \"instruction\": \"[Generate by Substrate]\",\n",
    "            \"input\": substrates_string_formatted,\n",
    "            \"output\": f\"Seq=<{train_sequences[i]}>\"\n",
    "        }\n",
    "        train_data_list.append(data_point_generate)\n",
    "\n",
    "        # 配列から基質を決定する指示\n",
    "        data_point_determine = {\n",
    "            \"instruction\": \"[Determine Substrate]\",\n",
    "            \"input\": f\"Seq=<{train_sequences[i]}>\",\n",
    "            \"output\": substrates_string_formatted  # 整形された文字列を使用\n",
    "        }\n",
    "        train_data_list.append(data_point_determine)\n",
    "\n",
    "    with open(join(save_dir, \"train_split.json\"), 'w') as f:\n",
    "        json.dump(train_data_list, f, indent=4)\n",
    "\n",
    "    # テストデータを作成して保存\n",
    "    test_data_list = []\n",
    "    for i in range(len(test_sequences)):\n",
    "        substrates_string = \", \".join(test_substrates[i])\n",
    "        substrates_string_formatted = f\"Substrate=<{substrates_string}>\"\n",
    "\n",
    "        # 基質から配列を生成する指示\n",
    "        data_point_generate = {\n",
    "            \"instruction\": \"[Generate by Substrate]\",\n",
    "            \"input\": substrates_string_formatted,\n",
    "            \"output\": f\"Seq=<{test_sequences[i]}>\"\n",
    "        }\n",
    "        test_data_list.append(data_point_generate)\n",
    "\n",
    "        # 配列から基質を決定する指示\n",
    "        data_point_determine = {\n",
    "            \"instruction\": \"[Determine Substrate]\",\n",
    "            \"input\": f\"Seq=<{test_sequences[i]}>\",\n",
    "            \"output\": substrates_string_formatted  # 整形された文字列を使用\n",
    "        }\n",
    "        test_data_list.append(data_point_determine)\n",
    "\n",
    "    with open(join(save_dir, \"test_split.json\"), 'w') as f:\n",
    "        json.dump(test_data_list, f, indent=4)\n",
    "\n",
    "# シーケンスと基質を取得\n",
    "sequences = df_reduced[\"Sequence\"].tolist()\n",
    "substrates = df_reduced[\"Substrates\"].tolist()\n",
    "\n",
    "# データを分割し、JSONファイルに保存\n",
    "save_to_json_split(sequences, substrates, test_size=0.1, random_state=0, save_dir=\"dataset\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **(2) tokenizerをロードしてparquet形式のデータに変換**\n",
    "\n",
    "-   トレーニングとテストの各データセットに対して、json形式→parquet形式に変換"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:building instruction dataset...\n",
      "WARNING:root:building instruction dataset...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train_dataset size: 192300\n",
      "train longest number of tokens: 888\n",
      "test_dataset size: 21368\n",
      "test longest number of tokens: 779\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating parquet from Arrow format: 100%|██████████| 193/193 [00:04<00:00, 47.12ba/s]\n",
      "Creating parquet from Arrow format: 100%|██████████| 22/22 [00:00<00:00, 39.69ba/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "107589976"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from build_instruction_dataset import tokenize_instruction_dataset\n",
    "from transformers import AutoTokenizer\n",
    "\n",
    "MAX_SEQ_LEN = 1024\n",
    "TRAIN_DATA_PATH = \"dataset/train_split.json\"\n",
    "TEST_DATA_PATH = \"dataset/test_split.json\"\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"GreatCaptainNemo/ProLLaMA_Stage_1\")\n",
    "\n",
    "train_dataset = tokenize_instruction_dataset(tokenizer=tokenizer, data_cache_dir=None, data_path=TRAIN_DATA_PATH, max_seq_length=MAX_SEQ_LEN, preprocessing_num_workers=4)\n",
    "test_dataset = tokenize_instruction_dataset(tokenizer=tokenizer, data_cache_dir=None, data_path=TEST_DATA_PATH, max_seq_length=MAX_SEQ_LEN, preprocessing_num_workers=4)\n",
    "\n",
    "print(f\"train_dataset size: {len(train_dataset)}\")\n",
    "print(f\"train longest number of tokens: {max([len(x) for x in train_dataset['input_ids']])}\")\n",
    "print(f\"test_dataset size: {len(test_dataset)}\")\n",
    "print(f\"test longest number of tokens: {max([len(x) for x in test_dataset['input_ids']])}\")\n",
    "\n",
    "train_dataset.to_parquet(f\"dataset/train-{MAX_SEQ_LEN}.parquet\")\n",
    "test_dataset.to_parquet(f\"dataset/test-{MAX_SEQ_LEN}.parquet\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "unsloth_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
